Added Mushroom spread configuration to world config

GrayR0ot opened the pull request:
This update introduces a mushroom configuration to the Paper world, providing users with the ability to bypass the standard vanilla spread limit. This customization allows users to exert more control over mushroom spreading, potentially leading to increased mushroom growth within the designated area.

GrayR0ot committed:
Added Mushroom spread configuration to world config

GrayR0ot committed:
Added Mushroom spread configuration consumer to the minecraft MushroomBlock randomTick method

GrayR0ot committed:
Git patch UTF8 name fix

Spottedleaf committed:
Remove UpgradeData neighbour ticks outside of range

The lists are only supposed to contain ticks for the 1 radius
neighbours of the chunk.

lynxplay commented on patches/server/0005-Paper-config-files.patch:
```diff
@@ -1837,6 +1837,12 @@ index 0000000000000000000000000000000000000000..a33de97340f14219291c4175e9194914
++        public Mushroom mushroom;
++
++        public class Mushroom extends ConfigurationPart {
++            public int spreadLimit = 5;
```
As stated on the other PR, this should probably be a IntOr.Default.
We generally want to move away from magic values (like -1 for disabled or the 5 being defined in here)

GrayR0ot replied:
I tryed, but I can't use any IDE with Paper, I have a gradle issue with my IDEA. I looked for examples, but I haven't found.

GrayR0ot replied:
-1 mean unlimited, not disabled. In fact if the player let -1, it will calculate the maximum theorical value. Which is 9 x 9 x 3 in this case. It's better than letting the player calculate himself.

Source: https://minecraft.fandom.com/wiki/Mushroom#Spread

"When receiving a [random tick](https://minecraft.fandom.com/wiki/Tick#Random_tick), a mushroom has a chance of spreading to a nearby block (including even above and below diagonally) if there are fewer than five mushrooms of the same type in a **9×9×3 volume** around the original mushroom."

GrayR0ot replied:
[95c4e17](https://github.com/PaperMC/Paper/pull/9592/commits/95c4e17fe71bc8bdcdb391ae5cc4be363672deaf)

lynxplay replied:
Wouldn't "disabling'" a spread limit imply it to be, well, not be a factor? If my spread limit is the "default" I expect it to be vanilla, if I set it to 3, I expect it to be 3. If I disable it I expect it to not be enforced, allowing it to spread as far as possible.

GrayR0ot replied:
I can add a simple setting like "disable-limit" which would ignore the spread-limit option, and then set it to the maximum (9 x 9 x 3). What do you think?

lynxplay replied:
@Machine-Maker do you think it is worth looking into some addition IntOr to allow for both "disabled" and "default" here ?

GrayR0ot replied:
No, because if the player put -1, like a few other paper/spigot settings it's disabled by default. In my case this will works. If the player put -1 or 1000 or 1M it will disable the limit.

GrayR0ot replied:
But it still possible if needed for more clarity

lynxplay replied:
As said before, I think it is worth looking into the ability to move away from the magic constant that is -1 simply for clarity reasons.
We introduced this originally when world heights allowed for negative values, -1 would work here for sure, given that the limit cannot be negative.
This is purely a discussion in regards to clarity to the end user, specifically if we want to stick to -1 here or if we want to look into potentially expanding our configuration API to allow for the usage of both "default" and "disabled" in a single variable to communicate the implication of `-1` in a proper word.

GrayR0ot replied:
Oh I understand, yes you can add something like that. But I think it's more simple for a user to understand a field to enable/disable, & a field with the value when enabled.

lynxplay replied:
We'll wait for some more input on this, beyond that, the PR generally looks good tho, I'll have to think about the naming of the config option, given that spread limit could also be interpreted as "the radius" of the spread.

GrayR0ot replied:
Yes sure. The limit is not really easy to understand. I made a lot of test, in fact the limit is almost useless.
Minecraft will check in a 9 x 9 x 3 radius if there are already 5+ mushrooms. But it will trigger for all of them (all mushrooms can spread). I was looking to add an option to change the radius, but it's useless with spreading recursivity.

lynxplay commented on patches/server/1000-Increase-vanilla-mushroom-spread-limit.patch:
```diff
@@ -0,0 +1,29 @@
+             Iterator iterator = BlockPos.betweenClosed(pos.offset(-4, -1, -4), pos.offset(4, 1, 4)).iterator();
+ 
++            // Paper start - Increase vanilla mushroom spread
++            int mushroomSpreadLimit = world.paperConfig().environment.mushroom.spreadLimit;
```
I don't see the point in moving this down here ? Keep the logic where the initial i was defined so the removal and addition are in the same spot.

GrayR0ot replied:
Fixed

Warriorrrr committed:
Make the cursor limit option apply to banners (#9554)

GrayR0ot committed:
Updated edition lines

GrayR0ot committed:
Added IntOr.Default to 5 (Vanilla setting)

GrayR0ot committed:
Added BooleanOrDefault to disable vanilla mushroom-spread-limit fixed to 5

Warriorrrr committed:
Cache map ids on item frames (#9584)

montlikadani committed:
Fix players are invisible after using setPlayerProfile (#9143)

Co-authored-by: Jake Potrebic <jake.m.potrebic@gmail.com>
Co-authored-by: Bjarne Koll <lynxplay101@gmail.com>

GrayR0ot committed:
Fixed EoF

GrayR0ot reviewed (commented):
Added IntOrDefault as asked

Machine-Maker committed:
Add restrict-player-reloot-time config (#7652)

NoahvdAa committed:
Fix custom statistic criteria creation (#9595)

Machine-Maker committed:
Bandaid fix for Effect (#9548)

Effect or LevelEvent needs to be replaced
but ideally after the enum PR has been merged
upstream. Until then, this test and these fixes
should address all the known issues with them

Nacioszeczek committed:
SculkCatalyst bloom API (#9466)

Warriorrrr committed:
Fix DEFAULT getEntitySpawnReason for /summon (#9480)

kennytv committed:
[ci skip] Update README

Warriorrrr commented on patches/server/1000-Increase-vanilla-mushroom-spread-limit.patch:
```diff
@@ -0,0 +1,29 @@
+     @Override
+     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
+         if (random.nextFloat() < (world.spigotConfig.mushroomModifier / (100.0f * 25))) { // Spigot - SPIGOT-7159: Better modifier resolution
+-            int i = 5;
++            // Paper start - Increase vanilla mushroom spread
++            int mushroomSpreadLimit = world.paperConfig().environment.mushroom.spreadLimit.or(5);
```
Leaving the original i = 5 and using the i as the default in the .or might look a bit cleaner, but generally looks pretty good.

GrayR0ot committed:
fixup! Added BooleanOrDefault to disable vanilla mushroom-spread-limit fixed to 5

GrayR0ot committed:
Change requested

GrayR0ot committed:
Change requested

GrayR0ot committed:
Mushroom spread limit

GrayR0ot committed:
Merge branch 'feat/mushroom-spread-config' of github.com:GrayR0ot/Paper into feat/mushroom-spread-config

# Conflicts:
#	patches/server/0005-Paper-config-files.patch
